### ISSUE TITLE
[web] Emojis take a few seconds to render on canvaskit

### ISSUE THREAD

**Issue by mariamhas** [flutter/flutter#76248]
I'm running my app on CanvasKit on Chrome Linux and what I'm seeing is this for a second:
![image](https://user-images.githubusercontent.com/7997154/107508434-4c5cf800-6ba1-11eb-93b4-8679ed76e4b9.png)

Afterwards I'm seeing this for a few seconds:
![image](https://user-images.githubusercontent.com/7997154/107508404-40713600-6ba1-11eb-82e4-42f37c2c987d.png)

And then I finally see this:
![image](https://user-images.githubusercontent.com/7997154/107508494-61d22200-6ba1-11eb-8e00-dfc5cf3b4175.png)

I don't understand because I definitely can see the emoji's on a normal page in Chrome, for example this just renders: 💌. But why doesn't it render normally in Flutter, and why does it have to download a font 'Noto Color Emoji' to render the emoji's. And this is only for CanvasKit, when using the HTML render it just shows me the emoji's properly without these 'transition' phases and without downloading any emoji font. Can't CanvasKit use the normal font which is used for everything else on my desktop environment or Chrome?

_Originally posted by @rbozan in https://github.com/flutter/flutter/issues/74823#issuecomment-776668247_

**Comment by rbozan** [comment]
I think this might be also related to https://github.com/flutter/flutter/issues/30700

**Comment by yjbanov** [comment]
The issue is that the fallback font we're using is huge. We should find a smaller font to fall back on.

> Can't CanvasKit use the normal font which is used for everything else on my desktop environment or Chrome?

Unfortunately not until fallback fonts are added to the web standards and implemented in browsers. Possibilities right now are: find a smaller font, or partition the emoji font into multiple smaller font files that are loaded on demand.

**Comment by devj3ns** [comment]
I am experiencing the same issue. 
I have a section where I just use one emoji in a text. When trying to scroll to the section the page freezes for one second before showing it.

When I remove the emoji it works fine.

**Comment by lucasege** [comment]
> Unfortunately not until fallback fonts are added to the web standards and implemented in browsers.
Sorry, do you have any idea how realistic this change would be? My assumption is to not count on it any time soon.

As of right now, CanvasKit appears unusable for any modern web app because the current fallback font doesn't support all emojis end users may be accustomed to (and it is very large, as you're aware). This may seem overdramatic, but it seems unacceptable to me to have to give users an excuse for why Emojis on our app look weird or don't work.

This is a shame since CanvasKit yields a better experience on nearly every other aspect.

I'm able to bundle the default system fonts (apple fonts, namely) with my project to get proper emojis, but these files are far too large (>200MB for Apple's emoji file).

Providing some way to access the system's fonts would be a huge help - that's the only thing keeping me from using CanvasKit at this point. If that's entirely impossible then that's quite a shame.

If you have any other suggestion to access system fonts in CanvasKit, please let me know. I'd love to resolve this.

**Comment by lukepighetti** [comment]
We're not only having the preloading `Noto Color Emoji` issue, but we're having an issue where a lazy loaded grid view of emojis is grinding the entire app to a halt during scroll. We used best practices to ensure that the Grid performance was good. This is an issue in build mode on safari web using CanvasKit renderer.

A separate issue would be the expectation that native emoji font would be used instead of `Noto Color Emoji`

In this screen recording the app has completely locked up. I am trying to scroll but it's totally frozen. If I wait 30 seconds it will resume.


https://user-images.githubusercontent.com/666539/113601488-677f3d80-960f-11eb-83e3-da4a3e52dfc7.mov

**Comment by lukepighetti** [comment]
Made a formalized issue for the above https://github.com/flutter/flutter/issues/79882

**Comment by yeasin50** [comment]
While this issue is hanging, and I just need 4emojis on my design,  I'm using images to handle UX.
> this isnt a solution
- photo-editor > pasting text emoji > increasing font size > save > load asset image on flutter.

**Comment by lonelyteapot** [comment]
How is this still open? This is a deal breaker for modern apps.

**Comment by vogtb** [comment]
I too am experiencing this issue. Just commenting to make sure this issue stays open :)

**Comment by mococa** [comment]
Please fix this. Pretty much every modern web app has emojis in it. It's not viable to download 9mb everytime in order to see them

**Comment by hamishjohnson** [comment]
Does anyone have a solution to at least to pre-load the emoji font so that users don't have to see a square box while it loads?

**Comment by Leffe108** [comment]
> Does anyone have a solution to at least to pre-load the emoji font so that users don't have to see a square box while it loads?

I use the below solution to pre-load the font when user lands on the home screen of my app. In my app, emoji is not used on this screen and it is safe to assume that it takes users at least one second to tap on anything to leave this screen. 

You may use this code without attributing me and alter it as you see fit. (eg. as public domain)

```
import 'package:flutter/widgets.dart';

/// After some delay this widget builds a Text() at opacity 0 that
/// contains an emoji to trigger loading of emoji font.
///
/// This widget however is not non-zero in size. So put it where it doesn't 
/// affect the layout. Eg below other sufficiently large content using a stack.
class LoadEmojiFont extends StatefulWidget {
  final Duration delay;

  const LoadEmojiFont({Key? key, this.delay = Duration.zero}) : super(key: key);

  @override
  State<LoadEmojiFont> createState() => _LoadEmojiFontState();
}

class _LoadEmojiFontState extends State<LoadEmojiFont> {
  bool loadEmoji = false;

  @override
  void initState() {
    () async {
      await Future<void>.delayed(this.widget.delay);
      this.setState(() {
        this.loadEmoji = true;
      });
    }();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return this.loadEmoji
        ? const Opacity(
            opacity: 0.0,
            child: Text('🏆'),
          )
        : Container();
  }
}
```

Then in a stack/row/column:
```
if (kIsWeb) const LoadEmojiFont(delay: Duration(seconds: 1)),
```

If you do need to show emojis on your first view, then you may need to dive deeper if the font can be pre-loaded during flutter load, or maybe delay that content a bit?

**Comment by yendu** [comment]
Any update ?

**Comment by mwoelk** [comment]
Please also be aware that loading on demand is not allowed in EU due to GDPR because any access to a *.gstatic.com domain would be in violation. This is due to IP adresses being deemed personal information which would be transmitted to a third party without the users consent. So ther should be some way to provide fallbacks or font files locally.

**Comment by n0ks** [comment]
If you only need a few emojis, you can customize your current font by adding new emojis using FontLab. 
This approach worked well for me. Since I only needed a handful of emojis, the font size increase was minimal.

**Comment by kevmoo** [comment]
We can recommend to use the following to speed up loading of google fonts on the web

```html
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
```

via https://fonts.google.com/selection/embed

**Comment by bldev2473** [comment]
Any Updates on this?
